## KNN - Digit-Recognizer
training set：第一行為label，其餘784行為pixel(28x28)

使用 KNN 算臨近5個training set是什麼數字，選擇最高票

資料來源 : https://www.kaggle.com/c/digit-recognizer/data

使用sklearn-KNN演算法完成kaggle的手寫辨識code如下：

In [8]:
# coding: utf-8
import csv
import sklearn.cross_validation as cross_validation
from sklearn import neighbors
from sklearn import neighbors
import sklearn.metrics as metrics

csvfile = open(r'C:\Users\User\Desktop\x-villageee\sklearn-KNN--Digit-Recognizer\train.csv', 'r')
reader = csv.reader(csvfile)
headers = next(reader)
#print (headers)

featureList = []
labelList = []
for row in reader:
    labelList.append(row[0])
    featureList.append(row[1:])
train_data, test_data, train_target, test_target = cross_validation.train_test_split(featureList,labelList, test_size=0.1, random_state=0)
 
knn = neighbors.KNeighborsClassifier()
knn.fit(train_data,train_target)
predict_test = knn.predict(test_data)
print (metrics.classification_report(test_target, predict_test))

f = open(r"C:\Users\User\Desktop\x-villageee\sklearn-KNN--Digit-Recognizer\result_0.1.csv", "w")
for i in predict_test:    
    f.write(i+'\n')
f.close()

             precision    recall  f1-score   support

          0       1.00      1.00      1.00       201
          1       0.96      0.99      0.98       246
          2       0.99      0.95      0.97       199
          3       0.96      0.94      0.95       212
          4       0.96      0.97      0.96       194
          5       0.93      0.95      0.94       181
          6       0.96      0.98      0.97       219
          7       0.91      0.97      0.94       238
          8       0.96      0.91      0.93       214
          9       0.94      0.90      0.92       196

avg / total       0.96      0.96      0.96      2100



由上report可知test data的準確率為96%，0的準確率最高，9的準確率最低

各label的precision、reall、F1-score 對應圖以 minitab軟體 繪圖後放在同repo(label.jpg)